ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,490 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+# cython: language_level=3
+
+import socket
+from libc.stdio cimport FILE, fflush, fclose
+from libc.stdlib cimport free
+from libc.string cimport memset
+from libc.stdint cimport uintptr_t
+from posix.stdio cimport open_memstream
+from cpython.ref cimport PyObject, Py_INCREF, Py_DECREF
+
+import logging
+from ucx_api_dep cimport *
+from ..exceptions import (
+    UCXError,
+    UCXConfigError,
+    UCXCanceled,
+)
+from .buffer_interface import get_buffer_data
+
+
+cdef assert_ucs_status(ucs_status_t status, msg_context=None):
+    if status != UCS_OK:
+        msg = "[%s] " % msg_context if msg_context is not None else ""
+        msg += ucs_status_string(status).decode("utf-8")
+        raise UCXError(msg)
+
+
+cdef ucp_config_t * _read_ucx_config(dict user_options) except *:
+    """
+    Reads the UCX config and returns a config handle,
+    which should freed using `ucp_config_release()`.
+    """
+    cdef ucp_config_t *config
+    cdef ucs_status_t status
+    status = ucp_config_read(NULL, NULL, &config)
+    if status != UCS_OK:
+        raise UCXConfigError(
+            "Couldn't read the UCX options: %s" %
+            ucs_status_string(status).decode("utf-8")
+        )
+
+    # Modify the UCX configuration options based on `config_dict`
+    for k, v in user_options.items():
+        status = ucp_config_modify(config, k.encode(), v.encode())
+        if status == UCS_ERR_NO_ELEM:
+            raise UCXConfigError("Option %s doesn't exist" % k)
+        elif status != UCS_OK:
+            msg = "Couldn't set option %s to %s: %s" % \
+                  (k, v, ucs_status_string(status).decode("utf-8"))
+            raise UCXConfigError(msg)
+    return config
+
+
+cdef ucx_config_to_dict(ucp_config_t *config):
+    """Returns a dict of a UCX config"""
+    cdef char *text
+    cdef size_t text_len
+    cdef FILE *text_fd = open_memstream(&text, &text_len)
+    assert(text_fd != NULL)
+    ret = {}
+    ucp_config_print(config, text_fd, NULL, UCS_CONFIG_PRINT_CONFIG)
+    fflush(text_fd)
+    cdef unicode py_text = text.decode()
+    for line in py_text.splitlines():
+        k, v = line.split("=")
+        k = k[len("UCX_"):]
+        ret[k] = v
+    fclose(text_fd)
+    free(text)
+    return ret
+
+
+def get_current_options():
+    """
+    Returns the current UCX options
+    if UCX were to be initialized now.
+    """
+    cdef ucp_config_t *config = _read_ucx_config({})
+    ret = ucx_config_to_dict(config)
+    ucp_config_release(config)
+    return ret
+
+
+def get_ucx_version():
+    cdef unsigned int a, b, c
+    ucp_get_version(&a, &b, &c)
+    return (a, b, c)
+
+
+cdef void _listener_callback(ucp_ep_h ep, void *args) except *:
```

Comment:
As I just recently learned, this can sometimes behave funny when asyncio is involved as it well check exception set by asyncio and raise. Doing `cdef` without `void` or `except *` fixes that.

Reply:
I have rewritten the error handling, we now use https://github.com/madsbk/ucx-py/blob/cython_refactoring/ucp/_libs/ucx_api.pyx#L25 to handle callback exceptions.

Reply:
That sounds like a nice solution. Thanks Mads! 😄